mbus-api/src/mbus_nats.rs

```diff
@@ -0,0 +1,330 @@
+use crate::{
+    common::{
+        constants::*,
+        errors::JetstreamError,
+        retry::{backoff_with_options, BackoffOptions},
+    },
+    Bus, NatsMessage,
+};
+use async_nats::{
+    jetstream::{
+        self,
+        consumer::{
+            push::{Config, Messages},
+            DeliverPolicy,
+        },
+        stream::Stream,
+        Context,
+    },
+    Client,
+};
+use async_trait::async_trait;
+use futures::StreamExt;
+use serde::{de::DeserializeOwned, Serialize};
+use std::{io::ErrorKind, marker::PhantomData, time::Duration};
+
+/// Result wrapper for Jetstream requests
+pub type BusResult<T> = Result<T, JetstreamError>;
+
+/// Initialise the Nats Message Bus
+pub async fn message_bus_init(server: &str) -> NatsMessageBus {
+    NatsMessageBus::new(server).await
+}
+
+/// Nats implementation of the Bus
+#[derive(Clone)]
+pub struct NatsMessageBus {
+    client: Client,
+    jetstream: Context,
+}
+
+impl NatsMessageBus {
+    /// Connect to nats server
+    pub async fn connect(server: &str) -> Client {
+        tracing::info!("Connecting to the nats server {}...", server);
+        // We retry in a loop until successful. Once connected the nats library will handle
+        // reconnections for us.
+        let options = BackoffOptions::new()
+            .with_init_delay(Duration::from_secs(5))
+            .with_cutoff(4)
+            .with_delay_step(Duration::from_secs(2))
+            .with_max_delay(Duration::from_secs(10));
+        let mut tries = 0;
+        let mut log_error = true;
+        loop {
+            match async_nats::ConnectOptions::new()
+                .event_callback(|event| async move {
+                    match event {
+                        async_nats::Event::Disconnected => {
+                            tracing::warn!("NATS connection has been lost")
+                        }
+                        async_nats::Event::Connected => {
+                            tracing::info!("NATS connection has been reestablished")
+                        }
+                        _ => (),
+                    }
+                })
+                .connect(server)
```

Comment:
is there a lazy connect available so we don't even have to handle initial connection?